### ISSUE TITLE
Are there plans to publicize this function?

### ISSUE THREAD

**Issue by dan-lugg** [square/kotlinpoet#892]
Some of the util-ish functions dealing with generating primitive literals, namely the one linked below, would IMO serve well as part of the public API.

https://github.com/square/kotlinpoet/blob/330c9b8562cdf3704a41278172ed02517ad1561b/kotlinpoet/src/main/java/com/squareup/kotlinpoet/Util.kt#L80

Perhaps under a `interface PrimitiveLiteralWriter` or similar.

Are there plans to expose this functionality as part of the API?

The multi-line string literals (generated by `%S` formatting for example) are obviously functional, but when the resulting code is to be human-edited, it can be a bit difficult to work with. The ability to generate single-line string literals would be ideal — whether as a different formatting placeholder (`%s`) or via an API function (`stringLiteralWithQuotes(s: String)` or similar)

**Comment by Egorand** [comment]
I'm not quite sure I understand the use case, can you please provide an example? If the issue is that `stringLiteralWithQuotes` produces a multiline string when a single-line one is expected, then I don't see how this will be solved by exposing the function in the API.

**Comment by dan-lugg** [comment]
I just realized that I completely fuddled this issue 😕 Perhaps this should be closed and a new one opened with the following intent:

In `stringLiteralWithQuotes`, there's a condition that branches to either use a multi-line string literal, with a call to `trimMargin()`:
https://github.com/square/kotlinpoet/blob/330c9b8562cdf3704a41278172ed02517ad1561b/kotlinpoet/src/main/java/com/squareup/kotlinpoet/Util.kt#L85

Or, branches to a non-multi-line string:
https://github.com/square/kotlinpoet/blob/330c9b8562cdf3704a41278172ed02517ad1561b/kotlinpoet/src/main/java/com/squareup/kotlinpoet/Util.kt#L111

It would be great to have these behaviors split into separate functions, `stringLiteralOnMultipleLines` and `stringLiteralOnSingleLine` for the first and second branches of that condition respectively.

When generating a string literal, the ability to toggle between these behaviors would be helpful, as I've found often the multi-line string literal output is less than ideal for code readability.

The placeholder `%S` produces single or multi, dependent on the input; `%Ss` for explicit single, and `%Sm` for explicit multi would allow the ability to control the desired output.

I had asked on StackOverflow about this; here's the relevant question for completeness: https://stackoverflow.com/q/60293241/409279

**Comment by Egorand** [comment]
In general, I'm personally against introducing new API to provide options for tweaking the style of generated code. There's an infinite number of things you might want to tweak so KotlinPoet prefers to be opinionated about the code style, rather than providing a large API surface for all kinds of tweaks. Another argument is that generated code shouldn't be considered source code, since it's produced by a tool and not written by a programmer, hence we favor correctness over readability, even though we make an effort to generate readable code. I'll let other contributors weigh in, but I don't think we should be introducing this change.

**Comment by JakeWharton** [comment]
Agreed.

Can you provide an example that looks bad?

**Comment by dan-lugg** [comment]
@JakeWharton Really any case where there's alot of newlines in the input string. I'm toying with a templating language that produces (or produces, compiles, and executes) classes in Kotlin, where the static template text content is wrapped in calls to `Appendable::appendln`, in the same vein as Microsoft's T4 (https://docs.microsoft.com/en-us/visualstudio/modeling/code-generation-and-t4-text-templates?view=vs-2019). When there's a number of newlines in the static text segments, the argument to `appendln` makes the resulting code hard to follow. However, that said...

@Egorand I can understand that position, where opinionated matters are ... opinionated. I'll withdraw this for that reason, and pursue a different strategy.

Nevertheless, great library; and I'm glad this exists for a whole bunch of reasons.

**Comment by Egorand** [comment]
Thanks @dan-lugg!

**Comment by JakeWharton** [comment]
We generate some big strings in SQL Delight. Aside from the broken indentation, it looks okay.

```kotlin
  private inner class QueryTerm<out T : Any>(
    private val value: String,
    mapper: (SqlCursor) -> T
  ) : Query<T>(queryTerm, mapper) {
    override fun execute(): SqlCursor = driver.executeQuery(1686746880, """
    |SELECT item.*
    |FROM item_index
    |JOIN item ON (docid = item.id)
    |WHERE content LIKE '%' || ?1 || '%' ESCAPE '\'
    |ORDER BY
    |  -- deprecated classes are always last
    |  deprecated ASC,
    |  CASE
    |    -- exact match
    |    WHEN className LIKE ?1 ESCAPE '\' THEN 1
    |    -- prefix match with no nested type
    |    WHEN className LIKE ?1 || '%' ESCAPE '\' AND instr(className, '.') = 0 THEN 2
    |    -- exact match on nested type
    |    WHEN className LIKE '%.' || ?1 ESCAPE '\' THEN 3
    |    -- prefix match (allowing nested types)
    |    WHEN className LIKE ?1 || '%' ESCAPE '\' THEN 4
    |    -- prefix match on nested type
    |    WHEN className LIKE '%.' || ?1 || '%' ESCAPE '\' THEN 5
    |    -- infix match
    |    ELSE 6
    |  END ASC,
    |  -- prefer "closer" matches based on length
    |  length(className) ASC,
    |  -- alphabetize to eliminate any remaining non-determinism
    |  packageName ASC,
    |  className ASC
    |LIMIT 50
    """.trimMargin(), 1) {
      bindString(1, value)
    }

    override fun toString(): String = "Item.sq:queryTerm"
  }
```
I'd love to see an actual example if you have one.

**Comment by dan-lugg** [comment]
@JakeWharton Sure, I'm on other tasks for the next few days (the issue is in a pet project of mine) but I'll reply back here with something soon.

Your example does read fine, the problem (as I see it) is when you have several multi-line string literals interspersed with control flow blocks.
